saleor/graphql/meta/types.py

```diff
@@ -90,34 +91,48 @@ class ObjectWithMetadata(graphene.Interface):
 
     @staticmethod
     def resolve_metadata(root: ModelWithMetadata, _info):
+        root = get_valid_metadata_instance(root)
         return resolve_metadata(root.metadata)
 
     @staticmethod
     def resolve_metafield(root: ModelWithMetadata, _info, *, key: str):
+        root = get_valid_metadata_instance(root)
         return root.metadata.get(key)
 
     @staticmethod
     def resolve_metafields(root: ModelWithMetadata, _info, *, keys=None):
+        root = get_valid_metadata_instance(root)
         return _filter_metadata(root.metadata, keys)
 
     @staticmethod
     def resolve_private_metadata(root: ModelWithMetadata, info):
+        root = get_valid_metadata_instance(root)
         return resolve_private_metadata(root, info)
 
     @staticmethod
     def resolve_private_metafield(root: ModelWithMetadata, info, *, key: str):
         check_private_metadata_privilege(root, info)
+        root = get_valid_metadata_instance(root)
         return root.private_metadata.get(key)
 
     @staticmethod
     def resolve_private_metafields(root: ModelWithMetadata, info, *, keys=None):
         check_private_metadata_privilege(root, info)
+        root = get_valid_metadata_instance(root)
         return _filter_metadata(root.private_metadata, keys)
 
     @classmethod
     def resolve_type(cls, instance: ModelWithMetadata, _info):
         if isinstance(instance, ChannelContext):
             # Return instance for types that use ChannelContext
             instance = instance.node
+        else:
+            instance = get_valid_metadata_instance(instance)
         item_type, _ = resolve_object_with_metadata_type(instance)
         return item_type
+
+
+def get_valid_metadata_instance(instance):
+    if isinstance(instance, Checkout):
+        instance = instance.metadata_storage
```

Comment:
Shouldn't we use dataloader here?

Reply:
Maybe instead of adding logic to all base metadata resolvers, we should override it on `Checkout` type(with dataloader).

Reply:
Good idea, it should be enough if we just load proper model in the Checkout type, while this could be kept unchanged.